Add Certificate Pinning 📌

p4checo opened the pull request:
### Issue

Closes #125 

### Changes

- Updated `Network.AuthenticationChallengeValidatorClosure` to a `AuthenticationChallengeHandler` protocol.
- Created `ServerTrustEvaluator`, conforming to `AuthenticationChallengeHandler` which performs public key pinning using Base64 encoded SHA256 strings of the certificate's `SubjectPublicKeyInfo` (SPKI).
- Created extension on `SecCertificate` to extract public key data
- Created extension on `Data` to calculate the hash from a certificate's public key data.
- Added dummy `CCommonCrypto` module to be able to use `CommonCrypto` on the project, as well as aBuild phase to check if the `/usr/include/CommonCrypto/CommonCryoto.h` header is available, and an entry on README.md explaining the situation.
- Fixed some new compilation errors on Xcode 9.3.1
- Did some improvements on UT's static resource loading
- Added Keychain test target (and dummy host app), since iOS 9 public key extraction code path requires accessing the Keychain, which to be accessible on UT's requires a host app.

p4checo commented on Sources/Network/AuthenticationChallengeHandler.swift:
```diff
@@ -0,0 +1,14 @@
+//
+//  AuthenticationChallengeValidator.swift
```
Ahaha 😅 True!

p4checo commented on Sources/Network/Pinning/PublicKeyAlgorithm.swift:
```diff
@@ -0,0 +1,71 @@
+    case ecDsaSecp384r1
+    case ecDsaSecp521r1
+
+    public static let allValues: [PublicKeyAlgorithm] = [.rsa2048,
```
Great stuff! Wasn't aware of that upcoming change in Swift 4.2. Thanks! 👍

p4checo commented on Sources/Network/Pinning/ServerTrustEvaluator.swift:
```diff
@@ -0,0 +1,381 @@
+public final class ServerTrustEvaluator {
+
+    // tag used to add and find the public key in the Keychain
+    static let keychainPublicKeyTag = "com.mindera.alicerce.ServerTrustEvaluator.keychainPublicKeyTag"
```
Excellent point!

However, I think we should be safe, because for iOS apps to share Keychain entries they have to explicitly belong to a group. [source](https://developer.apple.com/documentation/security/keychain_services/keychains)

> In iOS, apps have access to a single keychain (which logically encompasses the iCloud keychain). This keychain is automatically unlocked when the user unlocks the device and then locked when the device is locked. An app can access only its own keychain items, or those shared with a group to which the app belongs. It can't manage the keychain container itself.

Besides, for a key to be accessible in a group it has to be defined using the `kSecAttrAccessGroup`, which we're not defining.

Lastly, the need for using the Keychain is only required for iOS 9 (to retrieve public key data). Once we move to iOS 10+ we can remove all this Keychain code 😄

p4checo commented on Sources/Network/Pinning/ServerTrustEvaluator.swift:
```diff
@@ -0,0 +1,225 @@
+
+        let certificateChainLength = SecTrustGetCertificateCount(trust)
+
+        for i in configuration.certificateCheckingOrder.indices(forChainLength: certificateChainLength) {
```
Fixed 😉

p4checo commented on Sources/Persistence/DiskMemoryPersistenceStack.swift:
```diff
@@ -83,7 +83,7 @@ public final class DiskMemoryPersistenceStack: NSObject, PersistenceStack {
 
     // needs to be atomic since `NSCache` operations are not always made on the same queue/thread, i.e. caller
-    private let usedMemorySize = Atomic<UInt64>(0) // Size in bytes
+    fileprivate let usedMemorySize = Atomic<UInt64>(0) // Size in bytes
```
The compiler started complaining about it, not sure why. 🤷‍♂️ 

It shouldn't be an error, since on Swift 4 `private` properties should be accessible on extensions of the same type as long as they're on the same file...

Any idea?

renatorodrigues replied:
Looking at the source, I don't see any problem. Swift 4 still requires you to use `fileprivate` sometimes but it does when you use nested types. Weird. 🤔

p4checo replied:
It's weird indeed. This error only began showing up on Xcode 9.3.1, so maybe it's a regression on the compiler, or they backed on that "rule" 🤷‍♂️

portellaa replied:
weird i can't reproduce the problem. i'm using xcode 9.3.1, i have it as private and it works 🤔 weird

p4checo commented on Tests/AlicerceTests/Network/Pinning/SecCertificate+PublicKeyTestCase.swift:
```diff
@@ -0,0 +1,183 @@
+
+class SecCertificate_PublicKeyTestCase: XCTestCase {
+    
+    override func setUp() {
```
🙈

p4checo commented on Tests/AlicerceTests/Network/Pinning/Certificates/CertificateGenerationSteps.md:
```diff
@@ -0,0 +1,98 @@
+# Create OpenSSL Certificates
```
Agree, even though this is only useful if one wants to re-generate certificates in the future, which shouldn't happen since we "force" a constant certificate verify date when we create our test `SecTrust`'s 😉. 

I thought it would be nice to leave that info on the repo nonetheless because it took me quite some time to gather and get it right 😆, and might come in handy in the future.

renatorodrigues replied:
Indeed. It was a good call. 👏

p4checo commented on Tests/AlicerceTests/Utils.swift:
```diff
@@ -8,7 +8,9 @@
 import UIKit
 
-func dataFromFile(withBundleClass bundleClass: AnyClass, name: String, type: String) -> Data {
+final class TestDummy {}
```
Don't think so, because we use it on `CertificateUtils.swift` too

renatorodrigues replied:
You're right. For a moment I thought it was only used in this file. Nevermind then.

renatorodrigues commented on Sources/Network/Pinning/PublicKeyAlgorithm.swift:
```diff
@@ -0,0 +1,71 @@
+    case ecDsaSecp384r1
+    case ecDsaSecp521r1
+
+    public static let allCases: [PublicKeyAlgorithm] = [.rsa2048,
```
I agree with you but `allCases` will be the standard in Swift 4.2 [SE-0194 - Derived Collection of Enum Cases](https://github.com/apple/swift-evolution/blob/master/proposals/0194-derived-collection-of-enum-cases.md)

portellaa replied:
yeap saw that after the comment 😅 . Sorry for the spam. but `all` it's prettier 😭

portellaa commented:
@p4checo does it build on your machine? this was just a comment, no need to change, i believe this can be broken.

Is travis configured to use the latest xcode?

but leave it as it is, if this breaks the compiling. we could create an issue to remember us to fix this, no? just a thought 😊

p4checo commented:
@portellaa It also doesn't build on my machine 😓 

Travis is set up to use 9.3, not sure if it uses 9.3 or 9.3.1 underneath...

And yes, we can create an issue to revisit this further down the road 👍

p4checo commented:
@portellaa, @renatorodrigues Think I've found the issue 😅 

Somehow I think I changed the Swift version of Alicerce to use 3.3 on the project 🤦‍♂️ . 
When integrating via CocoaPods it worked fine, because it assumed 4.1 from the .podspec.

Fixed, and thanks for nudging me, which led me to look at it multiple times 🙏

portellaa commented:
Greaaat, well done 👏 🎉
